File: android/core/src/main/java/ru/maksimbulva/chess/core/engine/move/generator/PawnMoveGenerator.kt
package ru.maksimbulva.chess.core.engine.move.generator

import ru.maksimbulva.chess.core.engine.Piece
import ru.maksimbulva.chess.core.engine.Player
import ru.maksimbulva.chess.core.engine.board.Board
import ru.maksimbulva.chess.core.engine.board.Cell
import ru.maksimbulva.chess.core.engine.board.Vector2
import ru.maksimbulva.chess.core.engine.move.Move
import ru.maksimbulva.chess.core.engine.move.MoveBuilder
import ru.maksimbulva.chess.core.engine.otherPlayer
import ru.maksimbulva.chess.core.engine.position.Position
import kotlin.math.abs

internal class PawnMoveGenerator(
    private val player: Player
) : IMoveGenerator {

    private val initialRow = if (player == Player.Black) 6 else 1
    private val promotionRow = if (player == Player.Black) 0 else 7
    private val enPassantCaptureRow = if (player == Player.Black) 3 else 4

    private val deltaRow = if (player == Player.Black) -1 else 1
    private val moveForwardDelta = Vector2(deltaRow, 0)

    private val captureDeltas = arrayOf(
        Vector2(deltaRow, -1),
        Vector2(deltaRow, 1)
    )

    private val otherPlayer = player.otherPlayer()

    override fun generateMoves(fromCell: Cell, position: Position, moves: MutableList<Move>) {
        val board = position.board
        val moveBuilder = MoveBuilder(fromCell)

        val moveForwardCell = fromCell + moveForwardDelta
        if (moveForwardCell != null && board.isEmpty(moveForwardCell)) {
            if (moveForwardCell.row == promotionRow) {
                generatePromotions(moveBuilder.setToCell(moveForwardCell), moves)
            } else {
                moves.add(moveBuilder.setToCell(moveForwardCell).build())
            }

            if (fromCell.row == initialRow) {
                val moveForwardTwiceCell = moveForwardCell + moveForwardDelta
                if (moveForwardTwiceCell != null && board.isEmpty(moveForwardTwiceCell)) {
                    moves.add(moveBuilder.setToCell(moveForwardTwiceCell).build())
                }
            }
        }

        generateCaptures(fromCell, board, moves)

        if (position.enPassantCaptureColumn != null
            && (fromCell.row == enPassantCaptureRow)
            && abs(fromCell.column - position.enPassantCaptureColumn) == 1
        ) {
            val toCell = Cell(fromCell.row + deltaRow, position.enPassantCaptureColumn)
            moves.add(moveBuilder.setToCell(toCell).setAsEnPassantCapture().build())
        }
    }

    private fun generateCaptures(
        fromCell: Cell,
        board: Board,
        moves: MutableList<Move>
    ) {
        for (captureDelta in captureDeltas) {
            val captureCell = fromCell + captureDelta
            if (captureCell != null && !board.isEmpty(captureCell)) {
                val pieceAtCaptureCell = board.pieceAt(captureCell)
                if (pieceAtCaptureCell?.player == otherPlayer) {
                    val moveBuilder = MoveBuilder(fromCell)
                        .setToCell(captureCell)
                        .setAsCapture(pieceAtCaptureCell.piece)
                    if (captureCell.row == promotionRow) {
                        generatePromotions(moveBuilder, moves)
                    } else {
                        moves.add(moveBuilder.build())
                    }
                }
            }
        }
    }

    private fun generatePromotions(
        moveBuilder: MoveBuilder,
        moves: MutableList<Move>
    ) {
        for (promoteTo in PAWN_PROMOTIONS) {
            moves.add(moveBuilder.setPromoteTo(promoteTo).build())
        }
    }
}

private val PAWN_PROMOTIONS = arrayOf(
    Piece.Knight,
    Piece.Bishop,
    Piece.Rook,
    Piece.Queen
)

File: ChessApp/ChessEngine/Move/Generator/PawnMoveGenerator.cs
using ChessEngine.Board;
using System;
using System.Collections.Generic;
using static System.Math;

namespace ChessEngine.Move.Generator
{
    internal sealed class PawnMoveGenerator : IPieceMoveGenerator
    {
        private readonly Player player;
        private readonly int initialRow;
        private readonly int prePromotionRow;
        private readonly int deltaRow;
        private readonly int enPassantCaptureRow;

        public PawnMoveGenerator(Player player)
        {
            this.player = player;
            initialRow = GetInitialRow(player);
            prePromotionRow = GetPrePromotionRow(player);
            deltaRow = GetDeltaRow(player);
            enPassantCaptureRow = GetEnPassantCaptureRow(player);
        }

        public void AppendMoves(BoardSquare originSquare, Position.Position position, List<Move> moves)
        {
            var board = position.Board;
            var moveBuilder = new MoveBuilder(Piece.Pawn, originSquare);
            var moveAppender = GetMoveAppender(originSquare);

            var moveForwardSquare = new BoardSquare(originSquare.Row + deltaRow, originSquare.Column);
            if (board.IsEmpty(moveForwardSquare))
            {
                moveAppender(moveBuilder.SetDestSquare(moveForwardSquare), moves);
                if (originSquare.Row == initialRow)
                {
                    var moveDoubleForwardSquare = GetDoubleForwardMoveSquare(originSquare);
                    if (board.IsEmpty(moveDoubleForwardSquare))
                    {
                        moves.Add(moveBuilder
                            .SetDestSquare(moveDoubleForwardSquare)
                            .SetPawnDoubleMove()
                            .Build());
                    }
                }
            }

            if (originSquare.Column > 0)
            {
                AppendIfCanCapture(
                    new BoardSquare(originSquare.Row + deltaRow, originSquare.Column - 1),
                    moveBuilder,
                    board,
                    moves,
                    moveAppender);
            }
            if (originSquare.Column + 1 < Board.Board.ColumnCount)
            {
                AppendIfCanCapture(
                    new BoardSquare(originSquare.Row + deltaRow, originSquare.Column + 1),
                    moveBuilder,
                    board,
                    moves,
                    moveAppender);
            }

            if (originSquare.Row == enPassantCaptureRow &&
                position.EnPassantCaptureColumn.Exists(x => Abs(originSquare.Column - x) == 1))
            {
                var enPassantDstSquare = new BoardSquare(
                    originSquare.Row + deltaRow,
                    position.EnPassantCaptureColumn.ValueOr(0));
                moves.Add(moveBuilder
                    .SetDestSquare(enPassantDstSquare)
                    .SetEnPassantCapture()
                    .Build());
            }
        }

        private void AppendIfCanCapture(
            BoardSquare destSquare,
            MoveBuilder moveBuilder,
            Board.Board board,
            List<Move> moves,
            Action<MoveBuilder, List<Move>> moveAppender)
        {
            if (board.IsEmpty(destSquare))
            {
                return;
            }

            var pieceAtDestSquare = board.GetPieceAt(destSquare);
            if (pieceAtDestSquare.player != player)
            {
                moveAppender(
                    moveBuilder.SetDestSquare(destSquare).SetCapture(pieceAtDestSquare.piece),
                    moves);
            }
        }

        private BoardSquare GetDoubleForwardMoveSquare(BoardSquare originSquare)
        {
            return new BoardSquare(originSquare.Row + deltaRow * 2, originSquare.Column);
        }

        private Action<MoveBuilder, List<Move>> GetMoveAppender(BoardSquare square)
        {
            if (square.Row == prePromotionRow)
            {
                return AppendMoveWithPromotions;
            }
            else
            {
                return AppendMoveAsIs;
            }
        }

        private static int GetInitialRow(Player player) => player == Player.Black ? 6 : 1;

        private static int GetPrePromotionRow(Player player) => player == Player.Black ? 1 : 6;

        private static int GetDeltaRow(Player player) => player == Player.Black ? -1 : 1;

        private static int GetEnPassantCaptureRow(Player player)
        {
            return player == Player.Black ? 3 : 4;
        }

        private static void AppendMoveAsIs(MoveBuilder moveBuilder, List<Move> moves)
        {
            moves.Add(moveBuilder.Build());
        }

        private static void AppendMoveWithPromotions(MoveBuilder moveBuilder, List<Move> moves)
        {
            moves.Add(moveBuilder.SetPromotion(Piece.Queen).Build());
            moves.Add(moveBuilder.SetPromotion(Piece.Rook).Build());
            moves.Add(moveBuilder.SetPromotion(Piece.Bishop).Build());
            moves.Add(moveBuilder.SetPromotion(Piece.Knight).Build());
        }
    }
}
